sdks/java/core/src/main/java/org/apache/beam/sdk/util/ApiSurface.java

```diff
@@ -78,6 +92,231 @@
   private static final Logger LOG = LoggerFactory.getLogger(ApiSurface.class);
 
   /**
+   * {@link Matcher}s for use in {@link ApiSurface} related tests that aim to keep the public API
+   * conformant to a hard-coded policy by controlling what classes are allowed to be exposed by an
+   * API surface.
+   */
+  // based on previous code by @kennknowles and others.
+  public static class Matchers {
+
+    private static class ClassInPackage extends TypeSafeDiagnosingMatcher<Class<?>> {
+
+      private final String packageName;
+
+      private ClassInPackage(final String packageName) {
+        this.packageName = packageName;
+      }
+
+      @Override
+      public void describeTo(final Description description) {
+        description.appendText("Classes in package \"");
+        description.appendText(packageName);
+        description.appendText("\"");
+      }
+
+      @Override
+      protected boolean matchesSafely(final Class<?> clazz,
+                                      final Description mismatchDescription) {
+        return clazz.getName().startsWith(packageName + ".");
+      }
+    }
+
+    private static class SurfaceMatcher extends TypeSafeDiagnosingMatcher<ApiSurface> {
```

Comment:
This is more specifically a `ClassesInSurfaceMatcher`, in other words it is a `Matcher<ApiSurface>` defined to be the conjunction of a set of `Matcher<Class<?>>` applied to every class in the API surface.